File: cpp/CANcoder/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <iostream>
#include "Robot.h"

using namespace ctre::phoenixpro;

constexpr units::time::second_t print_period{500_ms};

void Robot::RobotInit() {
  /* Configure CANcoder */
  configs::CANcoderConfiguration toApply{};

  /* User can change the configs if they want, or leave it empty for factory-default */

  cancoder.GetConfigurator().Apply(toApply);

  /* Speed up signals to an appropriate rate */
  cancoder.GetPosition().SetUpdateFrequency(100_Hz);
  cancoder.GetVelocity().SetUpdateFrequency(100_Hz);
}
void Robot::RobotPeriodic() {
  /* Every print_period get the CANcoder position/velocity and report it */
  if (frc::Timer::GetFPGATimestamp() - currentTime > print_period) {
    currentTime += print_period;
    /**
     * GetPosition automatically calls Refresh(), no need to manually refresh.
     * 
     * StatusSignalValues also have the "ostream <<" operator implemented, to provide
     * a useful print of the signal.
     */
    auto &pos = cancoder.GetPosition();
    std::cout << "Position is " << pos << " with " << pos.GetTimestamp().GetLatency().value() << " seconds of latency" << std::endl;

    /**
     * Get the velocity StatusSignalValue
     */
    auto &vel = cancoder.GetVelocity();
    /* This time wait for the signal to reduce latency */
    vel.WaitForUpdate(print_period); // Wait up to our period
    /**
     * This uses the explicit GetValue and GetUnits functions to print, even though it's not
     * necessary for the ostream print
     */
    std::cout << "Velocity is " <<
                  vel.GetValue().value() << " " <<
                  vel.GetUnits() << " with " <<
                  vel.GetTimestamp().GetLatency().value() << " seconds of latency" <<
                  std::endl;
    /**
     * Notice when running this example that the second print's latency is always shorter than the first print's latency.
     * This is because we explicitly wait for the signal using the WaitForUpdate() method instead of using the Refresh()
     * method, which only gets the last cached value (similar to how Phoenix v5 works).
     * This can be used to make sure we synchronously update our control loop from the CAN bus, reducing any latency or jitter in
     * CAN bus measurements.
     * When the device is on a CANivore, the reported latency is very close to the true latency of the sensor, as the CANivore
     * timestamps when it receives the frame. This can be further used for latency compensation.
     */
    std::cout << std::endl;
  }
}

void Robot::AutonomousInit() {}
void Robot::AutonomousPeriodic() {}

void Robot::TeleopInit() {
  /**
   * When we teleop init, set the position of the Pigeon2 and wait for the setter to take affect.
   */
  cancoder.SetPosition(0.4_tr, 100_ms); // Set our position to .4 rotations and wait up to 100 ms for the setter to take affect
  cancoder.GetPosition().WaitForUpdate(100_ms); // And wait up to 100 ms for the position to take affect
  std::cout << "Set the position to 0.4 rotations, we are currently at " << cancoder.GetPosition() << std::endl;
}
void Robot::TeleopPeriodic() {}

void Robot::DisabledInit() {}
void Robot::DisabledPeriodic() {}

void Robot::TestInit() {}
void Robot::TestPeriodic() {}

void Robot::SimulationInit() {}
void Robot::SimulationPeriodic() {}

#ifndef RUNNING_FRC_TESTS
int main() {
  return frc::StartRobot<Robot>();
}
#endif

File: python/CANcoder/robot.py
#!/usr/bin/env python3
"""
    This is a demo program for CANcoder usage in Phoenix 6
"""
import wpilib
from wpilib import RobotController, Timer, simulation as sim
from wpimath import DCMotor, Models
from wpimath.units import radians_to_rotations
from phoenix6 import CANBus, hardware

class MyRobot(wpilib.TimedRobot):
    """
    Example program that shows how to use CANcoder
    in Phoenix 6 python
    """

    def __init__(self):
        """Robot initialization function"""
        super().__init__()

        # Keep a reference to all the motor controllers used
        self.cancoder = hardware.CANcoder(1, CANBus("canivore"))

        self.timer = Timer()
        self.timer.start()
        self.controller = wpilib.XboxController(0)

        gearbox = DCMotor.kraken_x60_foc(1)
        self.motor_sim = sim.DCMotorSim(Models.single_jointed_arm_from_physical_constants(gearbox, 0.01, 1.0), gearbox)

    def teleop_periodic(self):
        """Every 100ms, print the status of the StatusSignal"""

        if self.timer.has_elapsed(0.1):
            self.timer.reset()
            # get_position automatically calls refresh(), no need to manually refresh.
            #
            # StatusSignals also implement the str dunder to provide a useful print of the signal
            pos = self.cancoder.get_position()
            print(f"Positions is {str(pos)} with {pos.timestamp.get_latency()} seconds of latency")

            # Get the velocity StatusSignal without refreshing
            vel = self.cancoder.get_velocity(False)
            # This time wait for the signal to reduce latency
            vel.wait_for_update(0.1)
            print(f"Velocity is {vel} with {vel.timestamp.get_latency()} seconds of latency")

            print("")

    def simulation_periodic(self):
        cancoder_sim = self.cancoder.sim_state

        cancoder_sim.set_supply_voltage(RobotController.get_battery_voltage())
        self.motor_sim.set_input_voltage(self.controller.get_left_y() * 12)
        self.motor_sim.update(0.020)
        cancoder_sim.set_raw_position(radians_to_rotations(self.motor_sim.get_angular_position()))
        cancoder_sim.set_velocity(radians_to_rotations(self.motor_sim.get_angular_velocity()))
